Doc/c-api/monitoring.rst

```diff
@@ -0,0 +1,130 @@
+
+   Fire a ``LINE`` event.
+
+
+.. c:function:: int PyMonitoring_FireJumpEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *target_offset)
+
+   Fire a ``JUMP`` event.
+
+
+.. c:function:: int PyMonitoring_FireBranchEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *target_offset)
+
+   Fire a ``BRANCH`` event.
+
+
+.. c:function:: int PyMonitoring_FireCReturnEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *retval)
+
+   Fire a ``C_RETURN`` event.
+
+
+.. c:function:: int PyMonitoring_FirePyThrowEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *exception)
+
+   Fire a ``PY_THROW`` event.
+
+
+.. c:function:: int PyMonitoring_FireRaiseEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *exception)
+
+   Fire a ``RAISE`` event.
+
+
+.. c:function:: int PyMonitoring_FireCRaiseEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *exception)
+
+   Fire a ``C_RAISE`` event.
+
+
+.. c:function:: int PyMonitoring_FireReraiseEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *exception)
+
+   Fire a ``RERAISE`` event.
+
+
+.. c:function:: int PyMonitoring_FireExceptionHandledEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *exception)
+
+   Fire an ``EXCEPTION_HANDLED`` event.
+
+
+.. c:function:: int PyMonitoring_FirePyUnwindEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *exception)
+
+   Fire a ``PY_UNWIND`` event.
+
+
+.. c:function:: int PyMonitoring_FireStopIterationEvent(PyMonitoringState *state, PyObject *codelike, int32_t offset, PyObject *exception)
+
+   Fire a ``STOP_ITERATION`` event.
+
+
+Managing the Monitoring State
+-----------------------------
+
+Monitoring states can be managed with the help of monitoring scopes. A scope
+would typically correspond to a python function.
+
+.. :c:function:: int PyMonitoring_EnterScope(PyMonitoringState *state_array, uint64_t *version, const uint8_t *event_types, Py_ssize_t length)
+
+   Enter a monitoring scope. ``event_types`` is an array of the event IDs for
+   events that may be fired from the scope. For example, the ID of a ``PY_START``
+   event is the value ``PY_MONITORING_EVENT_PY_START``.
+   ``state_array`` is an array with a monitoring state entry for each event in
+   ``event_types``, it is allocated by the user but populated by
+   ``PyMonitoring_EnterScope`` with information about the activation state of
+   the event. The size of ``event_types`` (and hence also of ``state_array``)
+   is given in ``length``.
+
+   The ``version`` argument is a pointer to a value which should be initialized
+   to 0 and then set only by ``PyMonitoring_EnterScope`` itelf. It allows this
+   function to determine whether event states have changed since the previous call,
+   and to return quickly if they have not.
```

Comment:
If the scopes have the same set of events that they can fire, then I guess you could.

Reply:
Sorry, I'm still confused, and sadly I don't have enough time today to look through the code for answers. (Also, PEP 669 intentionally omits implementation details, and I don't know of a different write-up on the design.)

---

So, `state_array` and a corresponding `version` should be allocated and used together; they must be valid from the call `PyMonitoring_EnterScope` to the end of the corresponding `PyMonitoring_ExitScope`. Typically they are part of a CodeObject. Is that right?

Or can one have multiple scopes per “code object”? The Enter/Exit terminology makes me think of *frames* rather than *code objects*. If I emulate a code object for a recursive function, should `PyMonitoring_EnterScope` be called when the codelike is created, or when it's (first?) called?
(And if it's for every call: if I emulate a coroutine, should `PyMonitoring_EnterScope` be called every time it's resumed?)

---

The way the docs are written here, it looks like the user could pass a pointer to the same `static` variable every time.